// @auth0/nextjs-auth0/dist/auth0-session/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.clientFactory = exports.callbackHandler = exports.logoutHandler = exports.loginHandler = exports.getConfig = exports.TransientStore = exports.CookieStore = void 0;
var cookie_store_1 = require("./cookie-store");
Object.defineProperty(exports, "CookieStore", { enumerable: true, get: function () { return __importDefault(cookie_store_1).default; } });
var transient_store_1 = require("./transient-store");
Object.defineProperty(exports, "TransientStore", { enumerable: true, get: function () { return __importDefault(transient_store_1).default; } });
var get_config_1 = require("./get-config");
Object.defineProperty(exports, "getConfig", { enumerable: true, get: function () { return get_config_1.get; } });
var login_1 = require("./handlers/login");
Object.defineProperty(exports, "loginHandler", { enumerable: true, get: function () { return __importDefault(login_1).default; } });
var logout_1 = require("./handlers/logout");
Object.defineProperty(exports, "logoutHandler", { enumerable: true, get: function () { return __importDefault(logout_1).default; } });
var callback_1 = require("./handlers/callback");
Object.defineProperty(exports, "callbackHandler", { enumerable: true, get: function () { return __importDefault(callback_1).default; } });
var client_1 = require("./client");
Object.defineProperty(exports, "clientFactory", { enumerable: true, get: function () { return __importDefault(client_1).default; } });
