// @auth0/nextjs-auth0/dist/frontend/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.withPageAuthRequired = exports.useUser = exports.UserContext = exports.UserProvider = exports.useConfig = exports.ConfigProvider = void 0;
var use_config_1 = require("./use-config");
Object.defineProperty(exports, "ConfigProvider", { enumerable: true, get: function () { return __importDefault(use_config_1).default; } });
Object.defineProperty(exports, "useConfig", { enumerable: true, get: function () { return use_config_1.useConfig; } });
var use_user_1 = require("./use-user");
Object.defineProperty(exports, "UserProvider", { enumerable: true, get: function () { return __importDefault(use_user_1).default; } });
Object.defineProperty(exports, "UserContext", { enumerable: true, get: function () { return use_user_1.UserContext; } });
Object.defineProperty(exports, "useUser", { enumerable: true, get: function () { return use_user_1.useUser; } });
var with_page_auth_required_1 = require("./with-page-auth-required");
Object.defineProperty(exports, "withPageAuthRequired", { enumerable: true, get: function () { return __importDefault(with_page_auth_required_1).default; } });
